transport-native-epoll/src/main/c/netty_epoll_linuxsocket.c

```diff
@@ -262,6 +267,134 @@ static jobject netty_epoll_linuxsocket_getPeerCredentials(JNIEnv *env, jclass cl
      (*env)->SetIntArrayRegion(env, gids, 0, 1, (jint*) &credentials.gid);
      return (*env)->NewObject(env, peerCredentialsClass, peerCredentialsMethodId, credentials.pid, credentials.uid, gids);
 }
+
+static jint netty_epoll_linuxsocket_openTunTapCloneDevice0(JNIEnv* env, jclass clazz, jstring cloneDevName) {
+    const char* f_cloneDevName = (*env)->GetStringUTFChars(env, cloneDevName, 0);
+
+    // Open specific clone device.  Raise an IOException if an error occurs.
+    int res = open(f_cloneDevName, O_RDWR|O_NONBLOCK|O_CLOEXEC);
+    if (res < 0) {
+        netty_unix_errors_throwChannelExceptionErrorNo(env, "openTunTapCloneDevice() failed: ", errno);
+    }
+
+    (*env)->ReleaseStringUTFChars(env, cloneDevName, f_cloneDevName);
+
+    return res;
+}
+
+static jint netty_epoll_linuxsocket_allocTunTapInterface(JNIEnv *env, jclass cls, jint cloneDevFD, jstring ifName, jboolean isTapDev)
+{
+    jint res;
+    struct ifreq ifr;
+    const char *f_ifName = (*env)->GetStringUTFChars(env, ifName, NULL);
+
+    // Initialize the IFREQ structure with the specified flags and interface name.
+    memset(&ifr, 0, sizeof(ifr));
+    ifr.ifr_flags = ((isTapDev) ? IFF_TAP : IFF_TUN);
+    strncpy(ifr.ifr_name, f_ifName, IFNAMSIZ);
+
+    // Open/allocate the TUN/TAP device.  Raise an IOException if an error occurs.
+    res = ioctl(cloneDevFD, TUNSETIFF, (void *)&ifr);
+    if (res < 0) {
+        netty_unix_errors_throwChannelExceptionErrorNo(env, "allocTunTapInterface() failed: ", errno);
+    }
+
+    (*env)->ReleaseStringUTFChars(env, ifName, f_ifName);
+
+    return res;
+}
+
+static jlong netty_epoll_linuxsocket_readTunTapPacketAddress(JNIEnv *env, jclass cls, jint fd, jlong addr, jint pos, jint limit) {
+    jlong res = -1;
+    struct iovec iov[2];
+    struct tun_pi packetInfo;
+
+    iov[0].iov_base = (void *) &packetInfo;
+    iov[0].iov_len = sizeof(packetInfo);
+    iov[1].iov_base = ((char *) addr) + pos;
+    iov[1].iov_len = limit;
+
+    // Attempt to read a packet. Loop if the read is interrupted by a signal.
+    do {
+        res = (jlong) readv((int) fd, iov, 2);
+    } while (res < 0 && errno == EINTR);
+
+    // If a packet was received...
+    if (res >= (jlong) sizeof(packetInfo)) {
```

Comment:
what does it mean if exactly `packetInfo` was read? consider adding a comment to explain this case.